inc/c_util/tarray.h

```diff
@@ -0,0 +1,177 @@
+    if (tarray == NULL)                                                                                                                                         \
+    {                                                                                                                                                           \
+        /*Codes_SRS_TARRAY_02_005: [ If tarray is NULL then TARRAY_ENSURE_CAPACITY(T) shall fail and return a non-zero value. ]*/                               \
+        LogError("invalid argument TARRAY(" MU_TOSTRING(T) ") tarray=%p, uint32_t capacity=%" PRIu32 "",                                                        \
+            tarray, capacity);                                                                                                                                  \
+        result = MU_FAILURE;                                                                                                                                    \
+    }                                                                                                                                                           \
+    else                                                                                                                                                        \
+    {                                                                                                                                                           \
+        /*Codes_SRS_TARRAY_02_006: [ If tarray->capacity is greater than or equal to capacity then TARRAY_ENSURE_CAPACITY(T) shall succeed and return 0. ]*/    \
+        if (capacity <= tarray->capacity)                                                                                                                       \
+        {                                                                                                                                                       \
+            /*we already have the capacity*/                                                                                                                    \
+            result = 0;                                                                                                                                         \
+        }                                                                                                                                                       \
+        else                                                                                                                                                    \
+        {                                                                                                                                                       \
+            if(capacity > (((uint32_t)1)<<31))                                                                                                                  \
```

Comment:
>if(capacity > (((uint32_t)1)<<31))  [](http://example.com/codeflow?start=12&length=35)

Once could challenge that we can still grow here. Buuuut yeah, we can say these huge values are probably not of interest. At least we should document in the .md this limit. Maybe someone reads that. Maybe.

Reply:
This exists in order to avoid an overflow down the road where a multiplication result would extend past the UINT32_MAX.

However, it is now documented in .md as a separate spec.